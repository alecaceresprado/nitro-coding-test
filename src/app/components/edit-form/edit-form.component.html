<div class="edit-form">
  <div class="edit-form__title">Edit</div>
  <form [formGroup]="form" (ngSubmit)="onSubmit()">
    <div class="edit-form__controls">
      <div *ngFor="let field of editableFields" class="edit-form__field">
        <label class="edit-form__label" [for]="field">{{ field }}</label>
        <input class="edit-form__input" type="text" [formControlName]="field" />
        <p
          class="edit-form__error"
          *ngIf="
            !form.valid && form.dirty && submitted && !form.get(field).valid
          "
        >
          {{ field }} cannot be empty
        </p>
      </div>
    </div>
    <button class="edit-form__button" (click)="onSubmit()">Save changes</button>
  </form>
</div>
