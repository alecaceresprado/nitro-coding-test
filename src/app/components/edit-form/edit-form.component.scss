@import "../../../styles/vars";

.edit-form {
  @include layout();

  &__title {
    position: relative;
    background: white;
    top: -0.6rem;
    width: fit-content;
    left: 1rem;
    padding: 0 0.25rem;
  }

  &__field {
    padding: 1rem 1rem 0;
    &:last-child {
      padding-bottom: 1rem;
    }
  }

  &__label {
    padding-right: 0.5rem;
  }

  &__error {
    color: red;
    margin: 0.25rem 3rem;
  }

  &__button {
    position: relative;
    margin: 1rem;
    border: none;
    cursor: pointer;
    height: 1.5rem;
  }
}
