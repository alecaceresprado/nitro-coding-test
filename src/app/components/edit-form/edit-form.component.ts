import { Component, OnInit, Input, Output, EventEmitter } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { Store } from '@ngrx/store';

import { Post } from '@models';
import { AppState, updatePost } from '@state';

@Component({
  selector: 'app-edit-form',
  templateUrl: './edit-form.component.html',
  styleUrls: ['./edit-form.component.scss'],
})
export class EditFormComponent implements OnInit {
  @Input() public post: Post;
  @Output() public success: EventEmitter<void> = new EventEmitter();

  public form: FormGroup;
  public editing = false;
  public submitted = false;
  public editableFields: string[] = ['author', 'location'];

  constructor(private fb: FormBuilder, private store: Store<AppState>) {}

  ngOnInit(): void {
    this.form = this.fb.group(this.generateFields());
  }

  public onSubmit(): void {
    this.submitted = true;
    if (this.form.valid) {
      const payload = {
        id: this.post.id,
        author: '',
        location: '',
      };
      this.editableFields.forEach((field) => {
        payload[field] = this.form.get(field).value;
      });

      this.store.dispatch(updatePost(payload));
      this.success.emit();
      this.submitted = false;
    }
  }

  private generateFields() {
    return this.editableFields.reduce((result, current) => {
      result[current] = [this.post[current], Validators.required];
      return result;
    }, {});
  }
}
