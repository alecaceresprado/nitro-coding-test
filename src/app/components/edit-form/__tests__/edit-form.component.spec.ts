import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { provideMockStore } from '@ngrx/store/testing';
import { provideMockActions } from '@ngrx/effects/testing';

import { EditFormComponent } from '../edit-form.component';

const post = {
  id: 1,
  location: 'San Francisco',
  time: '1552657573',
  author: 'Happy User',
  text:
    'Proper PDF conversion ensures that every element of your document remains just as you left it.',
};

describe('EditFormComponent', () => {
  let component: EditFormComponent;
  let fixture: ComponentFixture<EditFormComponent>;
  let successFn = { emit: jest.fn() };

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [EditFormComponent],
      providers: [
        provideMockStore({ initialState: { posts: [] } }),
        provideMockActions({} as any),
      ],
      imports: [FormsModule, ReactiveFormsModule],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(EditFormComponent);
    component = fixture.componentInstance;
    component.post = post;
    component.success = successFn as any;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should render', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should handle submit', () => {
    component.onSubmit();
    expect(successFn.emit).toHaveBeenCalled();
  });
});
