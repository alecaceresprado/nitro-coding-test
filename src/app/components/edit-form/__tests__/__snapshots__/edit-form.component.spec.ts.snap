// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditFormComponent should render 1`] = `
<app-edit-form
  editableFields={[Function Array]}
  editing="false"
  fb={[Function FormBuilder]}
  form={[Function FormGroup]}
  post={[Function Object]}
  store={[Function MockStore]}
  submitted="false"
  success={[Function Object]}
>
  <div
    class="edit-form"
  >
    <div
      class="edit-form__title"
    >
      Edit
    </div>
    <form
      class="ng-untouched ng-pristine ng-valid"
      novalidate=""
    >
      <div
        class="edit-form__controls"
      >
        
        <div
          class="edit-form__field"
        >
          <label
            class="edit-form__label"
            for="author"
          >
            author
          </label>
          <input
            class="edit-form__input ng-untouched ng-pristine ng-valid"
            type="text"
          />
          
        </div>
        <div
          class="edit-form__field"
        >
          <label
            class="edit-form__label"
            for="location"
          >
            location
          </label>
          <input
            class="edit-form__input ng-untouched ng-pristine ng-valid"
            type="text"
          />
          
        </div>
      </div>
      <button
        class="edit-form__button"
      >
        Save changes
      </button>
    </form>
  </div>
</app-edit-form>
`;
